### ISSUE TITLE
prod mode failed: uglifyjs failed: SyntaxError: Unexpected token: name (Exception)

### ISSUE THREAD

**Issue by yanxiaodi** [driftyco/ionic-app-scripts#634]
**Note: for support questions, please use one of these channels:**

https://forum.ionicframework.com/
http://ionicworldwide.herokuapp.com/

#### Short description of the problem:
I have updated "@ionic/app-scripts" to 1.0.0, but the project can't be builded correctly. The build process throw the exception:

> E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp>ionic run browser --prod

> DS2@ ionic:build E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp
> ionic-app-scripts build "--prod"

[01:18:51]  ionic-app-scripts 1.0.0
[01:18:51]  build prod started ...
[01:18:51]  clean started ...
[01:18:52]  clean finished in 13 ms
[01:18:52]  copy started ...
[01:18:52]  ngc started ...
[01:19:08]  ngc finished in 16.41 s
[01:19:08]  webpack started ...
[01:19:09]  copy finished in 17.91 s
[01:19:19]  webpack finished in 11.34 s
[01:19:19]  uglifyjs started ...
[01:19:19]  sass started ...
[01:19:23]  sass finished in 3.54 s
[01:19:23]  cleancss started ...
[01:19:24]  uglifyjs failed: SyntaxError: Unexpected token: name (Exception)
[01:19:24]  ionic-app-script task: "build"
[01:19:24]  Error: SyntaxError: Unexpected token: name (Exception)
Error
    at new JS_Parse_Error (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@u
glify-js\tools\node.js:28:1), <anonymous>:1545:18)
    at js_error (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\t
ools\node.js:28:1), <anonymous>:1553:11)
    at croak (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\tool
s\node.js:28:1), <anonymous>:2092:9)
    at token_error (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-j
s\tools\node.js:28:1), <anonymous>:2100:9)
    at unexpected (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js
\tools\node.js:28:1), <anonymous>:2106:9)
    at semicolon (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\
tools\node.js:28:1), <anonymous>:2126:56)
    at simple_statement (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@ugl
ify-js\tools\node.js:28:1), <anonymous>:2317:73)
    at eval (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\tools
\node.js:28:1), <anonymous>:2186:19)
    at eval (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\tools
\node.js:28:1), <anonymous>:2139:24)
at block_ (eval at <anonymous> (E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.2.7.3@uglify-js\too
ls\node.js:28:1), <anonymous>:2432:20)

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\bin\\npm-c
li.js" "run" "ionic:build" "--" "--prod"
npm ERR! node v6.7.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! DS2@ ionic:build: `ionic-app-scripts build "--prod"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the DS2@ ionic:build script 'ionic-app-scripts build "--prod"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the DS2 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ionic-app-scripts build "--prod"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs DS2
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls DS2
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\npm-debug.log

#### What behavior are you expecting?
It's very strange because that the "ionic serve" command works fine. But the "ionic run browser --prod" or the "ionic build browser --prod" command failed with the "--prod" param. If there's no "--prod" param, it works as expected.

**Steps to reproduce:**
1. ionic run browser --prod. Or ionic build browser --prod.
2.
3.

```
insert any relevant code between the above and below backticks
```
This's my package.json:
{
    "name": "DS2",
    "author": "Loonyi",
    "homepage": "http://www.loonyi.com/",
    "private": true,
    "scripts": {
        "ionic:build": "ionic-app-scripts build",
        "ionic:serve": "ionic-app-scripts serve"
    },
    "dependencies": {
        "@angular/common": "2.2.1",
        "@angular/compiler": "2.2.1",
        "@angular/compiler-cli": "2.2.1",
        "@angular/core": "2.2.1",
        "@angular/forms": "2.2.1",
        "@angular/http": "2.2.1",
        "@angular/platform-browser": "2.2.1",
        "@angular/platform-browser-dynamic": "2.2.1",
        "@angular/platform-server": "2.2.1",
        "@ionic/storage": "1.1.7",
        "@types/echarts": "0.0.5",
        "echarts": "^3.3.2",
        "hi-base64": "^0.2.0",
        "ionic-angular": "2.0.0-rc.4",
        "ionic-native": "2.2.11",
        "ionicons": "3.0.0",
        "rxjs": "5.0.0-beta.12",
        "typescript-linq": "^1.0.6",
        "videojs-contrib-hls": "^3.6.14",
        "zone.js": "0.6.26"
    },
    "devDependencies": {
        "@ionic/app-scripts": "1.0.0",
        "sw-toolbox": "^3.4.0",
        "typescript": "2.0.9"
    },
    "description": "My project name",
    "cordovaPlugins": [
        "cordova-plugin-device",
        "cordova-plugin-console",
        "cordova-plugin-whitelist",
        "cordova-plugin-splashscreen",
        "cordova-plugin-statusbar",
        "ionic-plugin-keyboard"
    ],
    "cordovaPlatforms": []
}

**Which @ionic/app-scripts version are you using?**
"@ionic/app-scripts": "1.0.0",

**Other information:** (e.g. stacktraces, related issues, suggestions how to fix, stackoverflow links, forum links, etc)
The npm-debug.log:

0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files (x86)\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'ionic:build',
1 verbose cli   '--',
1 verbose cli   '--prod' ]
2 info using npm@3.10.3
3 info using node@v6.7.0
4 verbose run-script [ 'preionic:build', 'ionic:build', 'postionic:build' ]
5 info lifecycle DS2@~preionic:build: DS2@
6 silly lifecycle DS2@~preionic:build: no script for preionic:build, continuing
7 info lifecycle DS2@~ionic:build: DS2@
8 verbose lifecycle DS2@~ionic:build: unsafe-perm in lifecycle true
9 verbose lifecycle DS2@~ionic:build: PATH: C:\Program Files (x86)\nodejs\node_modules\npm\bin\node-gyp-bin;E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp\node_modules\.bin;C:\Program Files (x86)\nodejs;D:\Users\yxd\bin\Sencha\Cmd\5.1.3.61;C:\Program Files\Dell\DW WLAN Card;c:\Program Files (x86)\Intel\iCLS Client\;c:\Program Files\Intel\iCLS Client\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;c:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;c:\Program Files\Intel\Intel(R) Management Engine Components\DAL;c:\Program Files (x86)\Intel\Intel(R) Management Engine Components\IPT;c:\Program Files\Intel\Intel(R) Management Engine Components\IPT;C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;C:\Program Files (x86)\Microsoft SDKs\TypeScript\1.0\;C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\110\DTS\Binn\;C:\Program Files (x86)\Microsoft SQL Server\110\Tools\Binn\;C:\Program Files (x86)\Microsoft SQL Server\110\Tools\Binn\ManagementStudio\;C:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\IDE\PrivateAssemblies\;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn\;C:\Java\jdk1.8.0_101\bin;D:\Android\sdk;C:\Program Files (x86)\Skype\Phone\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files (x86)\Microsoft Emulator Manager\1.0\;C:\Program Files (x86)\nodejs\;c:\Program Files\Microsoft\Web Platform Installer\;C:\Program Files\Git\cmd;C:\Program Files\dotnet\;C:\Program Files (x86)\Microsoft VS Code\bin;C:\Users\yxd\AppData\Local\Microsoft\WindowsApps;C:\Users\yxd\AppData\Roaming\npm
10 verbose lifecycle DS2@~ionic:build: CWD: E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp
11 silly lifecycle DS2@~ionic:build: Args: [ '/d /s /c', 'ionic-app-scripts build "--prod"' ]
12 silly lifecycle DS2@~ionic:build: Returned: code: 1  signal: null
13 info lifecycle DS2@~ionic:build: Failed to exec ionic:build script
14 verbose stack Error: DS2@ ionic:build: `ionic-app-scripts build "--prod"`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (C:\Program Files (x86)\nodejs\node_modules\npm\lib\utils\lifecycle.js:242:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (C:\Program Files (x86)\nodejs\node_modules\npm\lib\utils\spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:877:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
15 verbose pkgid DS2@
16 verbose cwd E:\Source\Repos\DS2\Presentation\Loonyi.Site.WebApp
17 error Windows_NT 10.0.14393
18 error argv "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "ionic:build" "--" "--prod"
19 error node v6.7.0
20 error npm  v3.10.3
21 error code ELIFECYCLE
22 error DS2@ ionic:build: `ionic-app-scripts build "--prod"`
22 error Exit status 1
23 error Failed at the DS2@ ionic:build script 'ionic-app-scripts build "--prod"'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the DS2 package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     ionic-app-scripts build "--prod"
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs DS2
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls DS2
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]

I don't know is there anybody found this problem. Thanks!

**Comment by arendjantetteroo** [comment]
In my case it had to do with the fact that uglifyjs does not support ES6 syntax. https://github.com/mishoo/UglifyJS2/issues/448 and https://github.com/joeeames/WebpackFundamentalsCourse/issues/3

Uglifyjs is only run on prod builds, ionic serve does not use uglifyjs. 

It propably means your code is not correctly translated to ES5 by the current build process before it is handled to uglifyjs. I've not yet found the exact way to fix it in my case. 
Did it work in 0.0.48?

**Comment by yanxiaodi** [comment]
@arendjantetteroo Thanks for you reply!

I have found this method before I post here. And the target config in my tsconfig.json is "ES5". It's right.  So I created a new blank tabs project and it can build correctly. I add my codes step by step. At last I found the reason: I import a third party library named "typescript-linq". Maybe there're some codes caused this problem. The project build successfully after deleted this lib.

I found many problems like this :"uglifyjs failed: SyntaxError: Unexpected token: name (**some names**)". 

I think it's another main reason that some lib can't be supported by uglifyjs, except the reason you have replied. 

Thank you very much!
